text/3660-crates-io-crate-deletions.md

```diff
@@ -0,0 +1,123 @@
+- Start Date: 2024-06-20
+- RFC PR: [rust-lang/rfcs#3660](https://github.com/rust-lang/rfcs/pull/3660)
+
+# Summary
+
+This RFC proposes a mechanism for crate authors to delete their crates from crates.io under certain conditions.
+
+
+# Motivation
+
+There are a variety of reasons why a crate author might want to delete a crate or version from crates.io:
+
+* You published something accidentally.
+* You wanted to test crates.io.
+* You published content you didn't intend to be public.
+* You want to rename a crate. (The only way to rename a package is to re-publish it under a new name)
+
+The current [crates.io usage policy](https://crates.io/policies) says:
+
+> Crate deletion by their owners is not possible to keep the registry as immutable as possible.
+
+This restriction makes sense for the majority of crates that have been around for a while and are actively used, but the above list of reasons shows that there are valid use cases for allowing crate authors to delete their crates without having to contact the crates.io team.
+
+To make this process easier for our users and to reduce the workload of the crates.io team dealing with such support requests, we propose to codify our current set of informal rules into a formal policy that allows crate authors to delete their crates themselves under certain conditions (see below).
+
+
+# Proposal
+
+We propose to allow crate authors to delete their **crates** from crates.io under the following conditions:
+
+* The crate has been published for less than 72 hours,
+* or if all the following conditions are met:
+  * The crate has a single owner,
+  * The crate has no reverse dependencies,
+  * The crate has been downloaded less than 100 times for each month it has been published.
+
+We also propose to allow crate authors to delete **versions** of their crates from crates.io under the following conditions:
+
+* The version has been published for less than 72 hours.
```

Comment:
I feel like this will replace yank in a lot of situations and feel like more friction or guard rails are needed.

Reply:
This does seem worrisome to me. I think depending on a deleted crate would be very disruptive for those users. I'm uncertain of the exact behavior, but I think users of those crates will either get a confusing download error (if using Cargo.lock), or a confusing resolution error (if not using Cargo.lock).

I think it would be helpful to make sure the UI has loud and explicit warnings about the consequences of this.

Would it be possible to also warn owners of reverse dependencies when this happens?

Would it be possible to include a marker in the index similar to yanked so that cargo can provide a better error message?

Reply:
> Would it be possible to also warn owners of reverse dependencies when this happens?

Well, there aren't any owners of reverse dependencies, because "there are no reverse dependencies" is a requirement for deletion in this RFC.

I do think it's a good idea to include a marker in the index to provide a better error message for people who depend on the deleted crate in things which are not published to crates.io.

Reply:
> Well, there aren't any owners of reverse dependencies, because "there are no reverse dependencies" is a requirement for deletion in this RFC.

I'm responding specifically to the version deletion, which does allow reverse dependencies. IIUC, the "no reverse dependencies" is for entire crate deletions only.

Reply:
> I think users of those crates will either get a confusing download error (if using Cargo.lock)

I imagine it would be possible to store the SHA of anything removed, permitting cargo to provide a decent error message in this situation (i.e. downgrade/upgrade the dep)

Reply:
> I think it would be helpful to make sure the UI has loud and explicit warnings about the consequences of this.

I agree. This is certainly not an action that should be taken lightly. If implemented as a button on the website it should have a confirmation step that highlights the consequences of this action. If implemented as a cargo plugin or in cargo itself a similar confirmation step will probably make sense.

> Would it be possible to also warn owners of reverse dependencies when this happens?

- for crates that are deleted within 72 hours of being created: yes, that could be possible
- for crates that are deleted after 72 hours: in this case deletion is not possible if reverse deps exist
- for versions that are deleted: since most reverse deps are using implicit `^` version requirements the reverse dependency would fall back to the previous version, so I'm not sure if a warning is necessary in this case

> Would it be possible to include a marker in the index similar to yanked so that cargo can provide a better error message?

generally, yes. when I thought about this question yesterday I thought a `{ "version": "1.2.3", "deleted": true }` would be sufficient, but I have a feeling that older cargo versions won't like such an index record particularly much.

Reply:
> for versions that are deleted: since most reverse deps are using implicit ^ version requirements the reverse dependency would fall back to the previous version, so I'm not sure if a warning is necessary in this case

except this would break lockfiles.  Even without that, if someone has a "instant update" Dependabot action, their version requirement could also already be on the latest version, breaking them.

> generally, yes. when I thought about this question yesterday I thought a { "version": "1.2.3", "deleted": true } would be sufficient, but I have a feeling that older cargo versions won't like such an index record particularly much.

I wish I could remember why bumping the summary version is costly as that is an easy answer to this as older cargos would then ignore it.

Reply:
> summary version

what do you mean by that?

Reply:
> what do you mean by that?

I assume this means the `v` field, which allows updating how cargo interprets an index entry. The main problem with it is that older cargos silently ignore those entries which can be confusing to users. I think that is fine in this case.

Reply:
Versions older than 1.51 don't know about the `v` field and ignore the `v` field, and can fail in various ways depending on how the record was extended.

1.51 and newer will ignore the entire record if it is an unrecognized version.

Reply:
hmm, so if I understand correctly a `v: 3` with `deleted: true` isn't really helping if old cargo would essentially ignore the `v` and `deleted` fields 🤔 

I assume if we only included the `v`, `version` and `deleted` fields then old cargo would consider the whole index file invalid? or would it only consider these records invalid?

anyway, I'm open to adjusting the index to add the `deleted` information in some way if it helps cargo to display better errors and if we can figure out a way to make it work reasonably well with older cargo versions :)

Reply:
> I assume if we only included the v, version and deleted fields then old cargo would consider the whole index file invalid? or would it only consider these records invalid?

No, cargo parses line-by-line, so it would just skip over deleted versions.

I'm not personally too concerned about older cargos. I'm more concerned about current cargo's ability to display a reasonable message.

For the record, I still think this 72-hour policy is dangerous and am concerned about the consequences.

Reply:
@Turbo87 I noticed that this conversation was marked as resolved, but I didn't see a response or corresponding update to the RFC to resolve it. Would it be possible to at least add the index considerations to the unresolved section?

Reply:
> Would it be possible to at least add the index considerations to the unresolved section?

sure, I'll add that :)

Reply:
1. This was marked resolved without any commits being pushed
2. That still leaves "this is too lax".   Ideally this risk would be covered in the Drawbacks ("if we make the interface for deletion too easy and the metrics too light, people might over-rely on it rather than yanking") or at minimum left Unresolved.  I would very much want other crates.io reviewers to see this and weigh this when approving the RFC.